syntax = "proto3";

service RaftNode {
        rpc RequestVote(RequestVoteRequest) returns (ResultWithTerm);
        rpc AppendEntries(NodeArgs) returns (ResultWithTerm);
        rpc GetLeader(NoArgs) returns (LeaderResp);
        rpc Suspend(DurationArgs) returns (NoArgs);
}

message RequestVoteRequest {
        int32 term = 1;
        int32 candidateId = 2;
        int32 lastLogIndex = 3;
        int32 lastLogTerm = 4;
}

message NoArgs {}

message NodeArgs {
        int32 term = 1;
        int32 node_id = 2;
}

message ResultWithTerm {
        int32 term = 1;
        bool result = 2;
}

message DurationArgs {
        int32 duration = 1;
}

message LeaderResp {
        int32 leader_id = 1;
        string leader_addr = 2;
}
